openpgp-sdkms/src/lib.rs

```diff
@@ -0,0 +1,969 @@
+        Some("cv25519") => SupportedPkAlgo::Curve25519,
+        Some("nistp256") => SupportedPkAlgo::Ec(ApiCurve::NistP256),
+        Some("nistp384") => SupportedPkAlgo::Ec(ApiCurve::NistP384),
+        Some("nistp521") => SupportedPkAlgo::Ec(ApiCurve::NistP521),
+        _ => unreachable!("argument has a default value"),
+    };
+
+    let credentials = Credentials::new_from_env()?;
+    let http_client = credentials.http_client()?;
+
+    info!("create primary key");
+    let primary = PublicKey::create(
+        &http_client,
+        key_name.to_string(),
+        KeyRole::Primary,
+        &algorithm,
+    )
+    .context("could not create primary key")?;
+
+    info!("create decryption subkey");
+    let subkey = PublicKey::create(
+        &http_client,
+        key_name.to_string(),
+        KeyRole::Subkey,
+        &algorithm,
+    )?;
+
+    info!("generate certificate");
+    // Primary, UserID + sig, subkey + sig
+    let mut packets = Vec::<Packet>::with_capacity(5);
+
+    // Self-sign primary key
+    let prim: Key<PublicParts, PrimaryRole> = primary
+        .sequoia_key
+        .clone()
+        .expect("unloaded primary key")
+        .into();
+
+    packets.push(prim.into());
+
+    let mut prim_signer = SdkmsAgent::new_signer(&key_name)?;
+
+    let primary_flags = KeyFlags::empty().set_certification().set_signing();
+
+    let mut cert = Cert::try_from(packets)?;
+
+    // User ID + signature
+    let builder = SignatureBuilder::new(SignatureType::PositiveCertification)
+        .set_primary_userid(true)?
+        .set_key_flags(primary_flags)?
+        .set_preferred_hash_algorithms(vec![
+            HashAlgorithm::SHA512,
+            HashAlgorithm::SHA256,
+        ])?
+        .set_preferred_symmetric_algorithms(vec![
+            SymmetricAlgorithm::AES256,
+            SymmetricAlgorithm::AES128,
+        ])?;
+    let uid_sig = uid.bind(&mut prim_signer, &cert, builder)?;
+
+    cert = cert.insert_packets(vec![Packet::from(uid), uid_sig.into()])?;
+
+    // Subkey + signature
+    let subkey_public: Key<PublicParts, SubordinateRole> = subkey
+        .sequoia_key
+        .as_ref()
+        .expect("unloaded subkey")
+        .clone()
+        .into();
+
+    let subkey_flags = KeyFlags::empty()
+        .set_storage_encryption()
+        .set_transport_encryption();
+
+    let builder = SignatureBuilder::new(SignatureType::SubkeyBinding)
+        .set_hash_algo(HashAlgorithm::SHA512)
+        .set_key_flags(subkey_flags)?;
+
+    let signature = subkey_public.bind(&mut prim_signer, &cert, builder)?;
+    cert = cert
+        .insert_packets(vec![Packet::from(subkey_public), signature.into()])?;
+
+    info!("bind keys and store certificate in SDKMS");
+    let armored = String::from_utf8(cert.armored().to_vec()?)?;
+
+    let mut metadata = HashMap::<String, String>::new();
+    metadata.insert(SUBKEY.to_string(), subkey.uid()?.to_string());
+    metadata.insert("certificate".to_string(), armored);
```

Comment:
I'm thinking about how using the metadata system like this could go wrong. Have you considered creating a new json object, just for this crate? Eventually plenty of other crates could use metadata entries like this, resulting in name clashes, especially for entry names such as "certificate". Have you considered how to deal with future updates to sdkms. If I remember correctly, "subkey" may become a first class entry to the Sobject.

Reply:
Yes, we are doing a subkey link. But a JSON is the way to go, so we can also store the version and metadata, and avoid clashes Hadn't thought of it.